File: project/src/androidTest/java/com/expedia/bookings/test/phone/pagemodels/common/LaunchScreen.java
package com.expedia.bookings.test.phone.pagemodels.common;

import android.support.test.espresso.ViewInteraction;
import android.support.test.espresso.matcher.ViewMatchers;

import com.expedia.bookings.R;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.matcher.ViewMatchers.isDescendantOfA;
import static android.support.test.espresso.matcher.ViewMatchers.withEffectiveVisibility;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.allOf;

public class LaunchScreen {

	public static ViewInteraction hotelLaunchButton() {
		return onView(allOf(withId(R.id.hotels_button), withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE)));
	}

	public static ViewInteraction flightLaunchButton() {
		return onView(allOf(withId(R.id.flights_button), withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE)));
	}

	public static ViewInteraction carLaunchButton() {
		return onView(allOf(withId(R.id.cars_button), withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE)));
	}

	public static ViewInteraction lxLaunchButton() {
		return onView(allOf(withId(R.id.activities_button), withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE)));
	}

	public static ViewInteraction groundTransportLaunchButton() {
		return onView(allOf(withId(R.id.transport_button), withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE)));
	}

	public static ViewInteraction tripsButton() {
		return onView(withText(R.string.trips));
	}

	public static ViewInteraction shopButton() {
		return onView(withText(R.string.shop));
	}

	public static ViewInteraction accountButton() {
		return onView(withText(R.string.account_settings_menu_label));
	}

	public static void launchHotels() {
		hotelLaunchButton().perform(click());
	}

	public static void launchFlights() {
		flightLaunchButton().perform(click());
	}

	public static void launchCars() {
		carLaunchButton().perform(click());
	}

	public static void launchActivities() {
		lxLaunchButton().perform(click());
	}

	public static void launchGroundTransport() {
		groundTransportLaunchButton().perform(click());
	}

	public static void clickOnAirAttachBanner() {
		onView(withId(R.id.air_attach_banner)).perform(click());
	}

	public static ViewInteraction carLaunchButtonInDoubleRow() {
		return onView(allOf(withId(R.id.cars_button), isDescendantOfA(withId(R.id.double_row_lob_selector))));
	}

	public static ViewInteraction lxLaunchButtonInDoubleRow() {
		return onView(allOf(withId(R.id.activities_button), isDescendantOfA(withId(R.id.double_row_lob_selector))));
	}

	public static ViewInteraction carLaunchButtonInSingleRow() {
		return onView(allOf(withId(R.id.cars_button), isDescendantOfA(withId(R.id.lob_selector))));
	}

	public static ViewInteraction lxLaunchButtonInSingleRow() {
		return onView(allOf(withId(R.id.activities_button), isDescendantOfA(withId(R.id.lob_selector))));
	}

	public static ViewInteraction lobSingleRowWidget() {
		return onView(withId(R.id.lob_selector));
	}

	public static ViewInteraction lobDoubleRowWidget() {
		return onView(withId(R.id.double_row_lob_selector));
	}

	public static ViewInteraction fiveLOBDoubleRowWidget() {
		return onView(withId(R.id.double_row_five_lob_selector));
	}
}

File: project/src/androidTest/java/com/expedia/bookings/test/pagemodels/common/LaunchScreen.kt
package com.expedia.bookings.test.pagemodels.common

import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.ViewInteraction
import android.support.test.espresso.matcher.ViewMatchers.isCompletelyDisplayed
import android.support.test.espresso.matcher.ViewMatchers.isDescendantOfA
import android.support.test.espresso.matcher.ViewMatchers.withId
import android.support.test.espresso.matcher.ViewMatchers.withParent
import android.support.test.espresso.matcher.ViewMatchers.withText
import android.view.View
import com.expedia.bookings.BuildConfig
import com.expedia.bookings.R
import com.expedia.bookings.test.espresso.Common
import com.expedia.bookings.test.espresso.EspressoUtils
import org.hamcrest.Matchers.allOf
import java.util.concurrent.TimeUnit

object LaunchScreen {
    @JvmStatic val tripsButton = allOf(withText(R.string.trips), isDescendantOfA(withId(R.id.tab_layout)))
    @JvmStatic val shopButton = allOf(withText(R.string.shop), isDescendantOfA(withId(R.id.tab_layout)))
    @JvmStatic val accountButton = allOf(withText(R.string.account_settings_menu_label), isDescendantOfA(withId(R.id.tab_layout)))

    @JvmStatic fun waitForLOBHeaderToBeDisplayed() {
        EspressoUtils.waitForViewNotYetInLayoutToDisplay(allOf(withId(R.id.launch_lob_widget), withParent(withId(R.id.launch_list_widget))), 10, TimeUnit.SECONDS)
    }

    @JvmStatic fun hotelsLaunchButton(): ViewInteraction {
        waitForLOBHeaderToBeDisplayed()
        return onView(allOf<View>(withText("Hotels"), isCompletelyDisplayed()))
    }

    @JvmStatic fun flightLaunchButton(): ViewInteraction {
        waitForLOBHeaderToBeDisplayed()
        return onView(allOf<View>(withText("Flights"), isCompletelyDisplayed()))
    }

    @JvmStatic fun packagesLaunchButton(): ViewInteraction {
        waitForLOBHeaderToBeDisplayed()
        return onView(allOf<View>(withText(getPackagesLaunchButtonText(BuildConfig.brand)), isCompletelyDisplayed()))
    }

    @JvmStatic fun packagesLaunchButtonForPOS(pos: String): ViewInteraction {
        waitForLOBHeaderToBeDisplayed()
        return onView(allOf<View>(withText(getPackagesLaunchButtonTextForPOS(pos)), isCompletelyDisplayed()))
    }

    @JvmStatic fun activitiesLaunchButton(): ViewInteraction {
        Common.delay(1)
        waitForLOBHeaderToBeDisplayed()
        return onView(allOf<View>(withText("Things to Do"), isCompletelyDisplayed()))
    }

    @JvmStatic fun carsLaunchButton(): ViewInteraction {
        Common.delay(1)
        waitForLOBHeaderToBeDisplayed()
        return onView(allOf<View>(withText("Car Rentals"), isCompletelyDisplayed()))
    }

    @JvmStatic fun tripsButton(): ViewInteraction {
        return onView(tripsButton)
    }

    @JvmStatic fun shopButton(): ViewInteraction {
        return onView(shopButton)
    }

    @JvmStatic fun accountButton(): ViewInteraction {
        return onView(accountButton)
    }

    @JvmStatic fun getPackagesLaunchButtonText(brand: String): String {

        return when (brand) {
            "Orbitz", "CheapTickets" -> "Packages"
            "ebookers" -> "Flight + Hotel"
            "Travelocity" -> "Vacation Packages"
            else -> {
                "Bundle Deals"
            }
        }
    }

    @JvmStatic fun getPackagesLaunchButtonTextForPOS(pos: String): String {
        if (pos == "Australia" || pos == "New Zealand") {
            return "Hotel + Flight Deals"
        } else if (pos == "Canada") {
            return "Flight + Hotel"
        } else if (pos == "Japan" || pos == "Singapore" || pos == "Malaysia" || pos == "Thailand" || pos == "Germany" || pos == "France" || pos == "Italy" || pos == "South Korea" || pos == "Mexico") {
            return "Hotel + Flight"
        } else {
            return "Bundle Deals"
        }
    }
}
